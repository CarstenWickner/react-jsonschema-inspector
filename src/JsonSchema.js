import RefScope from "./RefScope";
import {
    isNonEmptyObject, listValues, mapObjectValues, mergeObjects
} from "./utils";

class JsonSchema {
    /**
     * Alternative factory method to the constructor, returning "undefined" if the given schema is not a non-empty object.
     *
     * @param {Object} rawSchema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {RefScope} scope collection of available $ref targets
     * @returns {JsonSchema|undefined}
     */
    static createIfNotEmpty(rawSchema, parserConfig, scope) {
        return isNonEmptyObject(rawSchema) ? new JsonSchema(rawSchema, parserConfig, scope) : undefined;
    }

    /**
     * Raw JSON Schema
     */
    schema;

    /**
     * Configuration steering how the JSON schema is being traversed/parsed
     */
    parserConfig;

    /**
     * @type RefScope
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {Object} schema the JSON Schema to represent
     * @param {Object} parserConfig configuration affecting how the JSON schema is being traversed/parsed
     * @param {?RefScope} scope collection of available $ref targets
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }

    getRelevantSchemaParts() {
        const { allOf, anyOf, oneOf } = this.schema;
        const includeAnyOf = anyOf && this.parserConfig && this.parserConfig.anyOf === "likeAllOf";
        const includeOneOf = oneOf && this.parserConfig && this.parserConfig.oneOf === "likeAllOf";
        if (allOf || includeAnyOf || includeOneOf) {
            let result = allOf;
            if (includeAnyOf) {
                result = result ? result.concat(anyOf) : anyOf;
            }
            if (includeOneOf) {
                result = result ? result.concat(oneOf) : oneOf;
            }
            return result;
        }
        return undefined;
    }

    /**
     * Look-up of values associated with a specific kind of field in the given JSON schema definition.
     * If the targeted schema is an array, the sub-schemas of its "items" will be ignored.
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {?Function} mergeFunction optional: how to combine two encountered values into one
     * @param {?Function} mappingFunction optional: converting the raw value before calling the mergeFunction
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getFieldValue(fieldName, mergeFunction = listValues, mappingFunction) {
        if (!isNonEmptyObject(this.schema)) {
            return undefined;
        }
        if (this.schema.$ref) {
            // schema variable is merely a reference to a separately defined schema
            const referencedSchema = this.scope.find(this.schema.$ref);
            // by convention, if $ref is specified, all other properties in the schema are being ignored
            return referencedSchema.getFieldValue(fieldName, mergeFunction, mappingFunction);
        }
        const rawValue = this.schema[fieldName];
        let value = mappingFunction ? mappingFunction(rawValue, this.parserConfig, this.scope) : rawValue;
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            // schema variable is supposed to be combined with a given list of sub-schemas
            value = schemaParts
                .map(part => new JsonSchema(part, this.parserConfig, this.scope).getFieldValue(fieldName, mergeFunction, mappingFunction))
                .reduce(mergeFunction, value);
        }
        return value;
    }

    /**
     * Version of the getFieldValue() function to look-up of nested JSON schema definitions (while preserving the appropriate reference scope).
     *
     * This method is intended to be used to look-up fields like "items" or "additionalItems".
     *
     * @param {String} fieldName name/key of the field to look-up
     * @param {Function} mergeFunction optional: how to combine two encountered values into one
     * @return {*} result of the given mergeFunction over all encountered values of fields with the given name
     */
    getSchemaFieldValue(fieldName, mergeFunction = listValues) {
        return this.getFieldValue(fieldName, mergeFunction, JsonSchema.createIfNotEmpty);
    }

    /**
     * Determine whether this schema represents an array and if so, return the sub-schema describing its contents.
     *
     * @returns {JsonSchema} type of array contents or null if it is not an array or its content type could not be determined
     */
    getTypeOfArrayItems() {
        return this.getSchemaFieldValue("items") || this.getSchemaFieldValue("additionalItems") || null;
    }

    /**
     * Look-up of any nested sub-schemas that may contain information about "properties" in the given JSON schema definition.
     * If the given JSON schema definition refers is an array, it is assumed to not have relevant information.
     * Instead, the referenced schemas of its "Items" will be returned.
     *
     * This method is intended to be used to look-up fields like "properties" or "required" which are not associated with arrays.
     *
     * BEWARE: if a schema consists of multiple sub-schemas and only one of them defines an "items" field, the other sibling schemas
     * will still be included in the result list - they are just assumed to not have any relevant fields we are looking for.
     *
     * @returns {Array<JsonSchema>} array listing sub-schemas that may define properties about their children
     */
    getPropertyParentSchemas() {
        if (!isNonEmptyObject(this.schema)) {
            return [];
        }
        if (this.schema.$ref) {
            const referencedSchema = this.scope.find(this.schema.$ref);
            // this schema is just a reference to another separately defined schema
            return referencedSchema.getPropertyParentSchemas();
        }
        const schemaParts = this.getRelevantSchemaParts();
        if (schemaParts) {
            let returnedSchemas = [this];
            schemaParts.forEach((rawSchemaPart) => {
                returnedSchemas = returnedSchemas.concat(new JsonSchema(rawSchemaPart, this.parserConfig, this.scope).getPropertyParentSchemas());
            });
            return returnedSchemas;
        }
        if (isNonEmptyObject(this.schema.items)) {
            // unsupported: specifying array of schemas referring to entries at respective positions in described array
            // schema.items contains a single schema, that specifies the type of any value in the described array
            return new JsonSchema(this.schema.items, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        if (isNonEmptyObject(this.schema.additionalItems)) {
            // the given schema is an array which may specify its content type in "additionalItems"
            return new JsonSchema(this.schema.additionalItems, this.parserConfig, this.scope).getPropertyParentSchemas();
        }
        // if there are no nested schemas, only this schema itself may contain properties
        return [this];
    }

    /**
     * Extract the properties mentioned in this schema â€“ also considering all nested sub-schema by using getPropertyParentSchemas().
     *
     * @returns {Object.<String, JsonSchema>} collection of all properties mentioned in this schema
     */
    getProperties() {
        return this.getPropertyParentSchemas().map(({ schema, scope }) => {
            // properties is an Object.<String, raw-json-schema>
            const { properties, required = [] } = schema;
            const rawProperties = Object.assign(
                {},
                ...required.map(value => ({ [value]: true })),
                isNonEmptyObject(properties) ? properties : {}
            );
            return mapObjectValues(rawProperties, rawSchema => new JsonSchema(rawSchema, this.parserConfig, scope));
        }).reduce(mergeObjects, {});
    }
}

export default JsonSchema;
