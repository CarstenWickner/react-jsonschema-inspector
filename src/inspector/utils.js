export function getNestedProperties(schema) {
    if (typeof schema !== 'object' || !schema) {
        return null;
    }
    if (schema.properties) {
        // if the given schema represents an "object", it will have a list of "properties"
        return schema.properties;
    }
    if (schema.items) {
        // if the given schema represents an "array", it will identify a schema for its "items"
        return getNestedProperties(schema.items);
    }
    // for other types, there are no nested items to show
    return null;
}

export function hasNestedProperties(schema) {
    if (typeof schema !== 'object' || !schema) {
        return false;
    }
    if (schema.properties) {
        // if the given schema represents an "object", it will have a list of "properties"
        return true;
    }
    if (schema.items) {
        // if the given schema represents an "array", it will identify a schema for its "items"
        return hasNestedProperties(schema.items);
    }
    // for other types, there are no nested items to show
    return false;
}