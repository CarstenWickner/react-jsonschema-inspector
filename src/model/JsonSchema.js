import RefScope from "./RefScope";

/**
 * Representation of a Json Schema, offering a number of convenient functions for traversing and extracting information.
 */
export default class JsonSchema {
    /**
     * Raw JSON Schema.
     *
     * @type {object}
     */
    schema;

    /**
     * Configuration steering how the json schema is being traversed/parsed.
     *
     * @type {{oneOf: ?{groupTitle: ?string}, anyOf: ?{groupTitle: ?string}}}
     */
    parserConfig;

    /**
     * @type {RefScope}
     */
    scope;

    /**
     * Constructor for a JsonSchema (wrapper).
     *
     * @param {object} schema - the JSON Schema to represent
     * @param {object} parserConfig - configuration affecting how the json schema is being traversed/parsed
     * @param {?RefScope} scope - collection of available $ref targets (will be generated based on `schema` if not provided)
     */
    constructor(schema, parserConfig, scope) {
        this.schema = schema;
        this.parserConfig = parserConfig;
        this.scope = scope || new RefScope(this);
    }
}
