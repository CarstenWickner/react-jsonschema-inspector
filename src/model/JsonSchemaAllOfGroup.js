import JsonSchemaGroup from "./JsonSchemaGroup";

export default class JsonSchemaAllOfGroup extends JsonSchemaGroup {
    // eslint-disable-next-line class-methods-use-this
    shouldBeTreatedLikeAllOf() {
        return true;
    }

    with(schemaOrGroup) {
        if (schemaOrGroup instanceof JsonSchemaGroup && schemaOrGroup.shouldBeTreatedLikeAllOf()) {
            // unwrap entries to avoid an AllOfGroup being nested in an AllOfGroup
            schemaOrGroup.entries.forEach(entry => this.with(entry));
            return this;
        }
        // other groups or single schemas should just be added to the entries array
        return super.with(schemaOrGroup);
    }

    getOptions() {
        return this.entries
            .filter(entry => entry instanceof JsonSchemaGroup)
            .map(nestedGroup => nestedGroup.getOptions());
    }
}
