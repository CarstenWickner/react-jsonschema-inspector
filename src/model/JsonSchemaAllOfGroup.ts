import JsonSchemaGroup from "./JsonSchemaGroup";

/**
 * Representation of an `allOf` element in a Json Schema.
 */
export default class JsonSchemaAllOfGroup extends JsonSchemaGroup {
    /**
     * Extension of method from super class for adding a given Json Schema or group to this group.
     *
     * @param {JsonSchema|JsonSchemaGroup} schemaOrGroup - entry to add to this group
     * @returns {JsonSchemaAllOfGroup} this (i.e. self-reference for chaining)
     */
    with(schemaOrGroup) {
        if (schemaOrGroup instanceof JsonSchemaGroup && schemaOrGroup.shouldTreatEntriesAsOne()) {
            // unwrap entries to avoid an AllOfGroup being nested in an AllOfGroup
            schemaOrGroup.entries.forEach((entry) => this.with(entry));
            return this;
        }
        // other groups or single schemas should just be added to the entries array
        return super.with(schemaOrGroup);
    }
}
