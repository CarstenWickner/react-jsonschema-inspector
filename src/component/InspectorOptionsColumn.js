import PropTypes from "prop-types";
import React from "react";
import classNames from "classnames";
import isDeepEqual from "lodash.isequal";

import InspectorItem from "./InspectorItem";
import { getColumnDataPropTypeShape } from "./renderDataUtils";

import { getIndexPermutationsForOptions } from "../model/schemaUtils";

const InspectorOptionsColumn = (props) => {
    const {
        options, contextGroup, selectedItem, filteredItems, trailingSelection, renderItemContent, onSelect
    } = props;
    return (
        <div
            className={classNames({
                "jsonschema-inspector-column": true,
                "with-selection": selectedItem,
                "trailing-selection": trailingSelection
            })}
            onClick={onSelect}
            role="presentation"
            tabIndex={-1}
        >
            {getIndexPermutationsForOptions(options).map(optionIndexes => (
                <InspectorItem
                    key={JSON.stringify(optionIndexes)}
                    identifier={`Option ${1 + optionIndexes[optionIndexes.length - 1]}`}
                    schemaGroup={contextGroup}
                    optionIndexes={optionIndexes}
                    selected={isDeepEqual(optionIndexes, selectedItem)}
                    matchesFilter={filteredItems ? filteredItems.some(filteredOption => isDeepEqual(filteredOption, optionIndexes)) : undefined}
                    onSelect={event => onSelect(event, optionIndexes)}
                    renderContent={renderItemContent}
                />
            ))}
        </div>
    );
};

const columnDataPropTypeShape = getColumnDataPropTypeShape(false);
InspectorOptionsColumn.propTypes = {
    options: columnDataPropTypeShape.options.isRequired,
    // eslint-disable-next-line react/require-default-props
    contextGroup: columnDataPropTypeShape.contextGroup,
    selectedItem: PropTypes.arrayOf(PropTypes.number),
    filteredItems: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
    trailingSelection: columnDataPropTypeShape.trailingSelection,
    onSelect: columnDataPropTypeShape.onSelect.isRequired,
    renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
};
InspectorOptionsColumn.defaultProps = {
    selectedItem: null,
    trailingSelection: false,
    filteredItems: null,
    renderItemContent: null
};

export default InspectorOptionsColumn;
