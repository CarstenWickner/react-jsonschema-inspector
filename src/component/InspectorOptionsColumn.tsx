import * as PropTypes from "prop-types";
import * as React from "react";
import classNames from "classnames";
import isDeepEqual from "lodash.isequal";

import { InspectorItem } from "./InspectorItem";
import { getColumnDataPropTypeShape } from "./renderDataUtils";
import { JsonSchemaGroup } from "../model/JsonSchemaGroup";
import { RenderItemContentFunction, RenderOptions, RenderColumnOnSelectFunction } from "../types/Inspector";

const columnDataPropTypeShape = getColumnDataPropTypeShape(true);

interface OptionsColumnDefaultProps {
    selectedItem: Array<number>,
    filteredItems: Array<Array<number>>,
    trailingSelection: boolean,
    renderItemContent: RenderItemContentFunction
}

interface OptionsColumnProps extends OptionsColumnDefaultProps {
    contextGroup: JsonSchemaGroup,
    options: RenderOptions,
    onSelect: RenderColumnOnSelectFunction
}

export class InspectorOptionsColumn extends React.Component<OptionsColumnProps> {
    static defaultOptionNameForIndex = (optionIndexes: Array<number>) => `Option ${optionIndexes.map((index) => index + 1).join("-")}`;

    renderSingleOption(optionIndexes: Array<number>, name: string) {
        const {
            contextGroup, selectedItem, filteredItems, renderItemContent, onSelect
        } = this.props;
        return (
            <InspectorItem
                name={name}
                schemaGroup={contextGroup}
                optionIndexes={optionIndexes}
                selected={isDeepEqual(optionIndexes, selectedItem)}
                matchesFilter={filteredItems ? filteredItems.some((filteredOption) => isDeepEqual(filteredOption, optionIndexes)) : undefined}
                onSelect={(event) => onSelect(event, optionIndexes)}
                renderContent={renderItemContent}
            />
        );
    }

    renderGroupOfOptions({
        groupTitle,
        options,
        optionNameForIndex = InspectorOptionsColumn.defaultOptionNameForIndex
    }: RenderOptions, parentOptionIndexes: Array<number> = []) {
        return (
            <>
                {groupTitle && (
                    <div
                        key="group-title"
                        className="optional-group-title"
                    >
                        <span>{groupTitle}</span>
                    </div>
                )}
                <ul key="list-of-options">
                    {options.map((optionOrNestedGroup, index) => {
                        const optionIndexes = parentOptionIndexes.concat([index]);
                        return (
                            <li key={JSON.stringify(optionIndexes)}>
                                {optionOrNestedGroup.options && this.renderGroupOfOptions(optionOrNestedGroup, optionIndexes)}
                                {!optionOrNestedGroup.options && this.renderSingleOption(optionIndexes, optionNameForIndex(optionIndexes))}
                            </li>
                        );
                    })}
                </ul>
            </>
        );
    }

    render() {
        const {
            options, selectedItem, trailingSelection, onSelect
        } = this.props;
        return (
            <div
                className={classNames({
                    "jsonschema-inspector-column": true,
                    "optional-groups": true,
                    "with-selection": selectedItem,
                    "trailing-selection": trailingSelection
                })}
                onClick={onSelect}
                role="presentation"
                tabIndex={-1}
            >
                {this.renderGroupOfOptions(options)}
            </div>
        );
    }

    static propTypes = {
        options: columnDataPropTypeShape.options.isRequired,
        // eslint-disable-next-line react/require-default-props
        contextGroup: columnDataPropTypeShape.contextGroup,
        selectedItem: PropTypes.arrayOf(PropTypes.number),
        filteredItems: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number)),
        trailingSelection: columnDataPropTypeShape.trailingSelection,
        onSelect: columnDataPropTypeShape.onSelect,
        renderItemContent: PropTypes.func // func({ string: name, boolean: hasNestedItems, boolean: selected, JsonSchema: schema })
    };

    static defaultProps: OptionsColumnDefaultProps = {
        selectedItem: null,
        trailingSelection: false,
        filteredItems: null,
        renderItemContent: null
    };
}
