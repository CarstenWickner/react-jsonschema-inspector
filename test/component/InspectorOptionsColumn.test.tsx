import * as React from "react";
import { shallow } from "enzyme";

import { InspectorOptionsColumn } from "../../src/component/InspectorOptionsColumn";

import { getOptionsInSchemaGroup } from "../../src/model/schemaUtils";
import { JsonSchema } from "../../src/model/JsonSchema";
import { JsonSchemaAllOfGroup } from "../../src/model/JsonSchemaAllOfGroup";
import { JsonSchemaAnyOfGroup, JsonSchemaOneOfGroup } from "../../src/model/JsonSchemaOptionalsGroup";
import { ParserConfig } from "../../src/types/ParserConfig";
import { isDefined } from "../../src/model/utils";

describe("renders correctly", () => {
    const oneOfOptionNameForIndex = (optionIndexes: Array<number>): string => `Exclusive Option ${optionIndexes.map((index) => index + 1).join("-")}`;
    const parserConfig: ParserConfig = {
        oneOf: { optionNameForIndex: oneOfOptionNameForIndex }
    };
    const contextGroup = new JsonSchemaAllOfGroup()
        .with(new JsonSchema({ description: "Foobar" }, {}))
        .with(
            new JsonSchemaOneOfGroup(parserConfig)
                .with(new JsonSchema({ title: "Foo" }, {}))
                .with(
                    new JsonSchemaAnyOfGroup(parserConfig).with(new JsonSchema({ title: "Bar" }, {})).with(new JsonSchema({ description: "Baz" }, {}))
                )
        );
    const options = getOptionsInSchemaGroup(contextGroup);
    it("with minimal/default props", () => {
        const component = shallow(<InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={(): void => {}} />);
        expect(component).toMatchSnapshot();
    });
    it("with selection", () => {
        const component = shallow(
            <InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={(): void => {}} selectedItem={[1, 0]} />
        );
        expect(component.hasClass("with-selection")).toBe(true);
        expect(component.hasClass("trailing-selection")).toBe(false);
        expect(component.find({ name: "Exclusive Option 1" }).prop("selected")).toBe(false);
        expect(component.find({ name: "Option 2-1" }).prop("selected")).toBe(true);
        expect(component.find({ name: "Option 2-2" }).prop("selected")).toBe(false);
        expect(component.find({ name: "Exclusive Option 1" }).prop("matchesFilter")).toBeUndefined();
        expect(component.find({ name: "Option 2-1" }).prop("matchesFilter")).toBeUndefined();
        expect(component.find({ name: "Option 2-2" }).prop("matchesFilter")).toBeUndefined();
    });
    it("with trailing selection", () => {
        const component = shallow(
            <InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={(): void => {}} selectedItem={[0]} trailingSelection />
        );
        expect(component.hasClass("with-selection")).toBe(true);
        expect(component.hasClass("trailing-selection")).toBe(true);
        expect(component.find({ name: "Exclusive Option 1" }).prop("selected")).toBe(true);
        expect(component.find({ name: "Option 2-1" }).prop("selected")).toBe(false);
        expect(component.find({ name: "Option 2-2" }).prop("selected")).toBe(false);
        expect(component.find({ name: "Exclusive Option 1" }).prop("matchesFilter")).toBeUndefined();
        expect(component.find({ name: "Option 2-1" }).prop("matchesFilter")).toBeUndefined();
        expect(component.find({ name: "Option 2-2" }).prop("matchesFilter")).toBeUndefined();
    });
    it("with filtered items", () => {
        const component = shallow(
            <InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={(): void => {}} filteredItems={[[1, 1]]} />
        );
        expect(component.hasClass("with-selection")).toBe(false);
        expect(component.hasClass("trailing-selection")).toBe(false);
        expect(component.find({ name: "Exclusive Option 1" }).prop("matchesFilter")).toBe(false);
        expect(component.find({ name: "Option 2-1" }).prop("matchesFilter")).toBe(false);
        expect(component.find({ name: "Option 2-2" }).prop("matchesFilter")).toBe(true);
    });
});
describe("calls onSelect", () => {
    const onSelect = jest.fn(() => {});
    const contextGroup = new JsonSchemaAllOfGroup()
        .with(new JsonSchema({ description: "Foobar" }, {}))
        .with(new JsonSchemaOneOfGroup({}).with(new JsonSchema({ title: "Foo" }, {})).with(new JsonSchema({ title: "Bar" }, {})));
    const options = {
        groupTitle: "one of",
        options: [{}, {}]
    };

    it("clearing selection when clicking on column", () => {
        const component = shallow(<InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={onSelect} />);
        const onClickListener = component.find(".jsonschema-inspector-column").prop("onClick");
        expect(onClickListener).toBeDefined();
        if (isDefined(onClickListener)) {
            onClickListener({} as React.MouseEvent);
        }
        expect(onSelect.mock.calls).toHaveLength(1);
        // expect no second parameter indicating selected item
        expect(onSelect.mock.calls[0]).toHaveLength(1);
    });
    it("setting selection when clicking on option", () => {
        const component = shallow(<InspectorOptionsColumn options={options} contextGroup={contextGroup} onSelect={onSelect} />);
        expect(component.find({ name: "Option 1" }).exists()).toBe(true);
        expect(component.find({ name: "Option 2" }).exists()).toBe(true);
        component.find({ name: "Option 1" }).prop("onSelect")(null);
        expect(onSelect.mock.calls).toHaveLength(1);
        // expect second parameter indicating selected option
        expect(onSelect.mock.calls[0]).toHaveLength(2);
        expect(onSelect.mock.calls[0][1]).toEqual([0]);
    });
});
