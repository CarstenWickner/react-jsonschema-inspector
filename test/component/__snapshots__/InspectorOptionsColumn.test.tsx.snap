// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with minimal/default props 1`] = `
<div
  className="jsonschema-inspector-column optional-groups"
  onClick={[Function]}
  role="presentation"
  tabIndex={-1}
>
  <div
    className="optional-group-title"
    key="group-title"
  >
    <span>
      one of
    </span>
  </div>
  <ul
    key="list-of-options"
  >
    <li
      key="[0]"
    >
      <Component
        name="Exclusive Option 1"
        onSelect={[Function]}
        optionIndexes={
          Array [
            0,
          ]
        }
        schemaGroup={
          JsonSchemaAllOfGroup {
            "entries": Array [
              JsonSchema {
                "parserConfig": Object {},
                "schema": Object {
                  "description": "Foobar",
                },
                "scope": RefScope {
                  "baseUri": null,
                  "externalRefs": Map {},
                  "internalRefs": Map {
                    "#" => [Circular],
                  },
                  "otherScopes": Array [],
                },
              },
              JsonSchemaOneOfGroup {
                "entries": Array [
                  JsonSchema {
                    "parserConfig": Object {},
                    "schema": Object {
                      "title": "Foo",
                    },
                    "scope": RefScope {
                      "baseUri": null,
                      "externalRefs": Map {},
                      "internalRefs": Map {
                        "#" => [Circular],
                      },
                      "otherScopes": Array [],
                    },
                  },
                  JsonSchemaAnyOfGroup {
                    "entries": Array [
                      JsonSchema {
                        "parserConfig": Object {},
                        "schema": Object {
                          "title": "Bar",
                        },
                        "scope": RefScope {
                          "baseUri": null,
                          "externalRefs": Map {},
                          "internalRefs": Map {
                            "#" => [Circular],
                          },
                          "otherScopes": Array [],
                        },
                      },
                      JsonSchema {
                        "parserConfig": Object {},
                        "schema": Object {
                          "description": "Baz",
                        },
                        "scope": RefScope {
                          "baseUri": null,
                          "externalRefs": Map {},
                          "internalRefs": Map {
                            "#" => [Circular],
                          },
                          "otherScopes": Array [],
                        },
                      },
                    ],
                    "settings": Object {
                      "groupTitle": "any of",
                    },
                  },
                ],
                "settings": Object {
                  "groupTitle": "one of",
                  "optionNameForIndex": [Function],
                },
              },
            ],
          }
        }
        selected={false}
      />
    </li>
    <li
      key="[1]"
    >
      <div
        className="optional-group-title"
        key="group-title"
      >
        <span>
          any of
        </span>
      </div>
      <ul
        key="list-of-options"
      >
        <li
          key="[1,0]"
        >
          <Component
            name="Option 2-1"
            onSelect={[Function]}
            optionIndexes={
              Array [
                1,
                0,
              ]
            }
            schemaGroup={
              JsonSchemaAllOfGroup {
                "entries": Array [
                  JsonSchema {
                    "parserConfig": Object {},
                    "schema": Object {
                      "description": "Foobar",
                    },
                    "scope": RefScope {
                      "baseUri": null,
                      "externalRefs": Map {},
                      "internalRefs": Map {
                        "#" => [Circular],
                      },
                      "otherScopes": Array [],
                    },
                  },
                  JsonSchemaOneOfGroup {
                    "entries": Array [
                      JsonSchema {
                        "parserConfig": Object {},
                        "schema": Object {
                          "title": "Foo",
                        },
                        "scope": RefScope {
                          "baseUri": null,
                          "externalRefs": Map {},
                          "internalRefs": Map {
                            "#" => [Circular],
                          },
                          "otherScopes": Array [],
                        },
                      },
                      JsonSchemaAnyOfGroup {
                        "entries": Array [
                          JsonSchema {
                            "parserConfig": Object {},
                            "schema": Object {
                              "title": "Bar",
                            },
                            "scope": RefScope {
                              "baseUri": null,
                              "externalRefs": Map {},
                              "internalRefs": Map {
                                "#" => [Circular],
                              },
                              "otherScopes": Array [],
                            },
                          },
                          JsonSchema {
                            "parserConfig": Object {},
                            "schema": Object {
                              "description": "Baz",
                            },
                            "scope": RefScope {
                              "baseUri": null,
                              "externalRefs": Map {},
                              "internalRefs": Map {
                                "#" => [Circular],
                              },
                              "otherScopes": Array [],
                            },
                          },
                        ],
                        "settings": Object {
                          "groupTitle": "any of",
                        },
                      },
                    ],
                    "settings": Object {
                      "groupTitle": "one of",
                      "optionNameForIndex": [Function],
                    },
                  },
                ],
              }
            }
            selected={false}
          />
        </li>
        <li
          key="[1,1]"
        >
          <Component
            name="Option 2-2"
            onSelect={[Function]}
            optionIndexes={
              Array [
                1,
                1,
              ]
            }
            schemaGroup={
              JsonSchemaAllOfGroup {
                "entries": Array [
                  JsonSchema {
                    "parserConfig": Object {},
                    "schema": Object {
                      "description": "Foobar",
                    },
                    "scope": RefScope {
                      "baseUri": null,
                      "externalRefs": Map {},
                      "internalRefs": Map {
                        "#" => [Circular],
                      },
                      "otherScopes": Array [],
                    },
                  },
                  JsonSchemaOneOfGroup {
                    "entries": Array [
                      JsonSchema {
                        "parserConfig": Object {},
                        "schema": Object {
                          "title": "Foo",
                        },
                        "scope": RefScope {
                          "baseUri": null,
                          "externalRefs": Map {},
                          "internalRefs": Map {
                            "#" => [Circular],
                          },
                          "otherScopes": Array [],
                        },
                      },
                      JsonSchemaAnyOfGroup {
                        "entries": Array [
                          JsonSchema {
                            "parserConfig": Object {},
                            "schema": Object {
                              "title": "Bar",
                            },
                            "scope": RefScope {
                              "baseUri": null,
                              "externalRefs": Map {},
                              "internalRefs": Map {
                                "#" => [Circular],
                              },
                              "otherScopes": Array [],
                            },
                          },
                          JsonSchema {
                            "parserConfig": Object {},
                            "schema": Object {
                              "description": "Baz",
                            },
                            "scope": RefScope {
                              "baseUri": null,
                              "externalRefs": Map {},
                              "internalRefs": Map {
                                "#" => [Circular],
                              },
                              "otherScopes": Array [],
                            },
                          },
                        ],
                        "settings": Object {
                          "groupTitle": "any of",
                        },
                      },
                    ],
                    "settings": Object {
                      "groupTitle": "one of",
                      "optionNameForIndex": [Function],
                    },
                  },
                ],
              }
            }
            selected={false}
          />
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
